### ISSUE TITLE
Cyclic inheritance issue with recent find-class hack

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#884]
Since 5366201c107af30fe04ed9f02f8aabe7328fa815
we've got a new problem: we can't redefine any class that was overriden.  In particular, this means that we can't modify a class and `C-c C-c` it, or even asdf-reload the file.  I also believe that this causes two successive calls to `define-configuration` over the same class to hang Nyxt (untested).

The reason is simple:

```lisp
(defclass-export gtk-buffer (buffer)
  ...)

(setf (find-class 'buffer) (find-class 'gtk-buffer))
```

After the above has been evaluated once, `buffer` now refers to the `gtk-buffer` class.  So if we try to evaluate the class definition of `gtk-buffer` again, it will lookup the superclass which is itself, thus causing the compiler to enter an infinite loop.

What shall we do?  Can't think of a solution right now, so reverting is one option.

@jmercouris

**Comment by jmercouris** [comment]
Hm, tricky tricky. Can we place an expression in place of buffer or no?

**Comment by jmercouris** [comment]
If we can place an expression there, then we can have it logically *not* try to extend itself in a loop.

**Comment by Ambrevar** [comment]
No we can't :(

**Comment by jmercouris** [comment]
Actually we can, since defclass-export is our own macro, we can do whatever we want, in a normal class we could not, but here we can.

**Comment by Ambrevar** [comment]
It seems that we can introduce an intermediary symbol to break the cycle, then rename the intermediary class to the desired result.

Example:

```lisp
(defclass foo () ())
(defclass bar () ())
(setf (find-class 'bar)
	(find-class 'foo))
(defclass foo-temp (bar) ())
(setf (find-class 'foo)
	(find-class 'foo-temp))
```

In the above, the final 'foo class inherits from the original 'foo class.
The main drawback is that it makes it a bit harder to inspect the hierarchy since these intermediary classes get intertwined.

To generate a temporary name, we need to create a macro around `defclass` (which we already have).  I suggest the name be prefixed with the actuall class name to help readability, e.g. `(gensym (string CLASS-SYM))`.

**Comment by jmercouris** [comment]
I think that could work!

**Comment by Ambrevar** [comment]
I came up with the following, seems to work:

```lisp
(defmacro defclass-export (name supers &body (slots . options))
  (if (and (find-class name nil)
           (member (find-class name)
                   (append (mapcar #'find-class supers)
                           (mapcar #'mopu:superclasses supers))))
      (let ((temp-name (gensym (string name))))
        ;; Don't export the class again.
        `(progn (defclass ,temp-name ,supers ,slots ,@options)
                (setf (find-class ',name) (find-class ',temp-name))))
      `(serapeum.exporting:defclass ,name ,supers ,slots ,@options)))
```

**Comment by jmercouris** [comment]
that's great to hear!

can you please update the macro in the source file?

**Comment by Ambrevar** [comment]
There are still something that's broken, with or without this macro.

Recipe:

- Load :nyxt/gtk and start Nyxt.
- Close it.
- Reload Nyxt.
- Start again: a condition is raised, about minibuffer-mode and a
  missing buffer.

I need to do more testing.

**Comment by Ambrevar** [comment]
I'm not very happy with defclass and its alternatives.  This cycle
issue is the cherry on top, now I think we really must roll our own
`define-class` macro.

I'm thinking of the following features:

- Implicit initarg, accessor, reader and writer like defclass/std.
- Implicit initform defaulting to the zero-value of the slot type
  (e.g. 0 for integers, "" for strings).
- New `:export' slot option.
- Global option to toggle the default slot exportation.
- Global option to change the initform default to (error "Slot must be set").
- Global option to change the name generation of the initarg, accessor,
  reader and writer.  `nil` means none no initarg, etc. is generated.
- Cycle detection in superclasses (like the above macro).

Example result:

```
(define-class foo (foo)
  ((slot1 :type integer) ; defaults to 0
   (slot2 :type string :initform "hello!")
   (unexported-slot :export nil)
  (:export-all t
   :initarg-names (lambda (s) (str:concat "my-" s)))

; And this would work
(make-instance 'foo :my-slot1 17)
```

Maybe have a `define-class-export` that exports the class name and turns
on slot exportation by default.

Such a macro would cut off hundreds of lines of code in Nyxt and possibly
fix the few typos that slipped under our radar :)

**Comment by jmercouris** [comment]
I agree with your conclusions, we need our own macro. We've long wished for a way to optionally specify a slot accessor to be automatically exported. It will also allow us to implement the cyclic check as you've described! There is no downside that I can see, because under it all are true classes, just an extension in their behavior at expansion time :-)

**Comment by jmercouris** [comment]
One point of contention for me is the initarg, I believe this should not be implicit. A unset slot value can be an indicator that some process has not completed, that an object has not been manipulated in some way, and this serves as a flag.

**Comment by Ambrevar** [comment]
John Mercouris <notifications@github.com> writes:

> One point of contention for me is the initarg, I believe this should not be implicit. A unset slot value can be an indicator that some process has not completed, that an object has not been manipulated in some way, and this serves as a flag.

Did you mean :initform?

Either way, we can tune the behaviour globally:

- Default to unbound.
- Default to type's zero value.
- Default to (error ...).

We can ever have globals to customize all classes defaults.
defclass/std does that.

**Comment by Ambrevar** [comment]
I've fixed master.  It should work as before now.
I'll fix compatibility with CCL and work on the define-class macro.

**Comment by Ambrevar** [comment]
I've studied hu.dwim.defclass*: It does much of what we want except for

- automatic initform (no `(error ...)` nor zero value of type);
- cycle detection.

So we need our own class nonetheless.  Question is, do we base our own
class on top of defclass* or do we roll our own?

defclass* works and it's well tested, but it's unmaintained, the code is
not very pretty (some very long functions), barely any documentation,
plus it drags a few dependencies (some hu.dwim libraries like
hu.dwim.asdf).

Thoughts?

**Comment by jmercouris** [comment]
Better to reinvent the wheel sometimes, we don't have to carry all the baggage of all the library. I think we would be best served with our own implementation.

**Comment by Ambrevar** [comment]
See #889.

I ended up using hu.dwim.defclass-star anyways because it works really
well and extending it was so easy, it saved me hours of work.

If we ever get stuck with it, we can always roll out our own.

**Comment by Ambrevar** [comment]
Fixed with #889.
